server/utils/challenge/challenge.utils.ts

```diff
@@ -47,31 +49,49 @@ export async function verifyChallenge({
   now,
   solution,
   publicKey: publicKeyParam,
-}: VerifyChallengeParams) {
+}: VerifyChallengeParams): Promise<null | string> {
   try {
     const [solutionVersion, solutionValue] = solution.split(".");
     const challengeClient = CHALLENGE_CLIENTS[solutionVersion];
 
     if (!challengeClient) {
-      throw new Error(`Invalid challenge version`);
+      return ErrorMessages.CHALLENGE_INVALID;
     }
 
+    // TODO: Make TTL for hash-based challenges shorter:
+
     const challengeTTL =
       isAnonChallenge(challenge) ||
       challenge.purpose === ChallengePurpose.SHARE_ROTATION
         ? Config.CHALLENGE_ROTATION_TTL_MS
         : Config.CHALLENGE_TTL_MS;
 
-    if (now - challenge.createdAt.getTime() >= challengeTTL) {
-      return false;
+    const challengeAge = now - challenge.createdAt.getTime();
+
+    const isChallengeAgeNearingTTL = challengeAge >= challengeTTL * 0.8;
+
+    if (process.env.NODE_ENV === "development") {
+      if (challengeAge >= challengeTTL) {
+        console.log(`❌ Challenge took ${ (challengeAge / 1000).toFixed(2) }s (> ${ (challengeTTL / 1000).toFixed(2) }s).`);
+
+        return `${ ErrorMessages.CHALLENGE_EXPIRED_ERROR } Took ${ (challengeAge / 1000).toFixed(2) }s (> ${ (challengeTTL / 1000).toFixed(2) }s).`;
+      } else if (isChallengeAgeNearingTTL) {
+        console.log(`⚠️ Challenge took ${ (challengeAge / 1000).toFixed(2) }s (~80% of ${ (challengeTTL / 1000).toFixed(2) }s).`);
+      } else {
+        console.log(`✅ Challenge took ${ (challengeAge / 1000).toFixed(2) }s (< ${ (challengeTTL / 1000).toFixed(2) }s).`);
+      }
+    } else {
+      console.log(`⚠️ Challenge for user ${ session.userId } took ${ (challengeAge / 1000).toFixed(2) }s (~80% of ${ (challengeTTL / 1000).toFixed(2) }s). userAgent = ${ session.userAgent }`);
```

Comment:
Here ~80% is hardcoded in this `console.log`. I think the percent should be calculated right ?